### PR TITLE
closest-feature

### PR DESCRIPTION
**Pull request by hjc1983**
enhancement on #55 about access parent node, this feature allows traverse upwards from the current element to find the first ancestor of DOM elements. 

cheers

### CODE REVIEW

**Comment by IonicaBizau** [comment]
Can you please write tests for it?
Thank you!

File: test/index.js [hunk 1]
```diff
@@ -109,6 +110,33 @@ tester.describe("scrape-it", t => {
...
+        scrapeIt(URL, {
+            addresses: {
+                listItem: "table tbody tr",
+                data: {
+                    address: ".address",
+                    suburb: {
+                        selector: ".city",
+                        closest: "table",
+                        convert(html) {
+                            $ = cheerio.load(html);
```
Comment (IonicaBizau) [inline test/index.js]: We can get rid of `cheerio` here, by passing the `$elm` variable as second argument: https://github.com/IonicaBizau/scrape-it/blob/810560e4729b6cc6a7414fa1871ff38bc6e9c143/lib/index.js#L167

It's going to be a performance improvement as well.

File: lib/index.js [hunk 2]
```diff
@@ -112,6 +119,7 @@ scrapeIt.scrapeHTML = ($, opts) => {
                 v.how = $elm => $elm.attr(v.attr);
             }
             objDef(v, "trimValue", true);
+            objDef(v, "closest", "", true);
```
Comment (IonicaBizau) [inline lib/index.js]: I think we don't need the `true` argument in this case.

File: lib/index.js [hunk 3]
```diff
@@ -164,7 +172,11 @@ scrapeIt.scrapeHTML = ($, opts) => {
                     }
 
                     if (cOpt.convert) {
-                        value = cOpt.convert(value);
+                        if (cOpt.closest) {
+                            var $node = $cContext.closest(cOpt.closest);
+                            value = cOpt.convert($node.html());
```
Comment (IonicaBizau) [inline lib/index.js]: So, here, we can pass the `$node` value in the second argument.

File: lib/index.js [hunk 4]
```diff
@@ -164,7 +172,11 @@ scrapeIt.scrapeHTML = ($, opts) => {
...
+                        } else
+                            value = cOpt.convert(value);
```
Comment (IonicaBizau) [inline lib/index.js]: And here the `$elm` in the second argument.

File: lib/index.js [hunk 5]
```diff
@@ -164,7 +172,11 @@ scrapeIt.scrapeHTML = ($, opts) => {
                     }
 
                     if (cOpt.convert) {
-                        value = cOpt.convert(value);
+                        if (cOpt.closest) {
+                            var $node = $cContext.closest(cOpt.closest);
+                            value = cOpt.convert($node.html());
+                        } else
```
Comment (IonicaBizau) [inline lib/index.js]: Please use curly braces here too. :grin:

**Comment by hjc1983** [comment]
thanks for the review, it was very useful. however, i'm unsure about remove _cheerio_ from Test, could you elaborate on that?

File: test/index.js (hunk 1 above)
Comment (IonicaBizau) [inline test/index.js]: @hjc1983 After your latest change, you'll receive the `$node` element in the second parameter of this `convert` function. Then you can simply do:

```js
convert (html, $node) {
  return $node.find(".city").text();
}
```

Comment (hjc1983) [inline test/index.js]: :+1: wow, this is cool

**Comment by IonicaBizau** [comment]
@hjc1983 Thanks once again! :+1:

**Pull request merged**